assignments/introductory-assignments.md

```diff
@@ -488,14 +582,767 @@ A digital circuit computer with 2GB of memory can distinguish 2^{17 179 869 184}
+
+The part of selfie that generates assembly is called a *disassembler*.
+
+[Disassembler](https://en.wikipedia.org/wiki/Disassembler) A computer program that translates machine language into assembly language - the inverse operation to that of an assembler.
+
+Selfie does not implement an assembler, though.
+The starc compiler generates machine code right away without going through assembly first (requiring an assembler to generate machine code from that assembly).
+
+First 10 lines of the `selfie.s` file:
+
+```
+0x0(~1): 0x000392B7: lui t0,0x39
+0x4(~1): 0xFC028293: addi t0,t0,-64
+0x8(~1): 0x00028193: addi gp,t0,0
+0xC(~1): 0x00000513: addi a0,zero,0
+0x10(~1): 0x0D600893: addi a7,zero,214
+0x14(~1): 0x00000073: ecall
+0x18(~1): 0x00750513: addi a0,a0,7
+0x1C(~1): 0x00800293: addi t0,zero,8
+0x20(~1): 0x025572B3: remu t0,a0,t0
+0x24(~1): 0x40550533: sub a0,a0,t0
+```
+
+What you see is a human-readable version of the machine code in `selfie.m`. The purpose of `selfie.s` is to study `selfie.m` and eventually understand its semantics.
+
+Each line represents one machine instruction.
+
+**Example**
+
+The fifth line `0x10(~1): 0x0D600893: addi a7,zero,214`, for example, reads like this.
+
+The hexadecimal number `0x10` is the 32-bit-word-aligned memory address of the instruction in memory.
+
+The expression `(~1)` is the approximate line number of the source code, in this case, `selfie.c`, compiled to this instruction.
+
+The 32-bit word `0x0D600893` is, in fact, the binary encoded version of the instruction itself.
+
+Finally, `addi $a7,$zero,214` is the instruction's human-readable assembly version.
+
+This means in particular that `0x0D600893` and `addi $a7,$zero,214` are semantically equivalent. 
+
+The 32-bit word `0x0D600893` in binary stored at address `0x10` in memory is thus the only thing that the machine needs; the rest is for us to make it readable.
+
+## RISC-V
+
+We know that `0x0D600893` represents `addi $a7,$zero,214` because it is specified precisely by the ISA of the open RISC-V processor family.
+
+RISC-V is an open-source architecture we use in this class.
+RISC-U is the RISC-V subset targeted, emulated, and virtualized by selfie.
+
+[RISC-V](https://en.wikipedia.org/wiki/RISC-V) an available instruction set architecture (ISA) based on established reduced instruction set computing (RISC) principles. In contrast to most ISAs, the RISC-V ISA can be freely used for any purpose, permitting anyone to design, manufacture, and sell RISC-V chips and software. While not the first open architecture ISA, it is significant because it is designed to be useful in a wide range of devices. The instruction set also has a substantial body of supporting software, which avoids new instruction sets' usual weakness.
```

Comment:
```suggestion
[RISC-V](https://en.wikipedia.org/wiki/RISC-V) an available instruction set architecture (ISA) based on established reduced instruction set computing (RISC) principles. In contrast to most ISAs, the RISC-V ISA can be freely used for any purpose, permitting anyone to design, manufacture, and sell RISC-V chips and software. While not the first open architecture, it is significant because it is designed to be useful in a wide range of devices. The instruction set also has a substantial supporting software body, which avoids new instruction sets' usual weakness.
```